ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/EventsTableTreeEvents.scala

```diff
@@ -175,7 +175,26 @@ private[events] trait EventsTableTreeEvents { this: EventsTable =>
       previousOffsetWhereClauseValues(startExclusive, previousEventNodeIndex)
     val witnessesWhereClause =
       sqlFunctions.arrayIntersectionWhereClause("tree_event_witnesses", requestingParty)
-    SQL"select #$selectColumns, array[$requestingParty] as event_witnesses, case when submitter = $requestingParty then command_id else '' end as command_id from participant_events where (event_offset > $startExclusive or (event_offset = $prevOffset and node_index > $prevNodeIndex)) and event_offset <= $endInclusive and #$witnessesWhereClause order by (#$orderByColumns) limit $pageSize"
+    SQL"""select #$selectColumns, array[$requestingParty] as event_witnesses, case when submitter = $requestingParty then command_id else '' end as command_id
+         from participant_events,
+          (select min(row_id) row_id from (
+            select min(row_id) row_id
+            from participant_events
+            where (event_offset > $startExclusive)
+              and event_offset <= $endInclusive
```

Comment:
Alternative significantly faster formulation found by @leo-da 

```suggestion
            select max(row_id + 1) row_id
            from participant_events
            where event_offset = $startExclusive
```

Reply:
`startExclusive` might not exist if it's beforeBegin, for example, but we can just add minimum possible row_id to the inner union here to account for that.

Reply:
```
 Aggregate  (cost=102.53..102.54 rows=1 width=4) (actual time=0.061..0.061 rows=1 loops=1)
   ->  Index Scan using participant_events_event_offset_idx on participant_events  (cost=0.43..102.41 rows=48 width=4) (actual time=0.024..0.046 rows=18 loops=1)
         Index Cond: (event_offset = '\x0000000000014fde'::bytea)
 Planning Time: 0.201 ms
 Execution Time: 0.084 ms
```
vs
```
 Result  (cost=0.53..0.54 rows=1 width=4) (actual time=31.867..31.867 rows=1 loops=1)
   InitPlan 1 (returns $0)
     ->  Limit  (cost=0.43..0.53 rows=1 width=4) (actual time=31.861..31.861 rows=1 loops=1)
           ->  Index Only Scan using participant_events_row_id_event_offset_node_index_idx on participant_events  (cost=0.43..44413.24 rows=430188 width=4) (actual time=31.859..31.859 rows=1 loops=1)
                 Index Cond: ((row_id IS NOT NULL) AND (event_offset > '\x0000000000014fde'::bytea))
                 Heap Fetches: 0
 Planning Time: 0.165 ms
 Execution Time: 31.893 ms
```

Reply:
I hypothesize that the reason for the mismatch between cost and actual results is that the planner thinks the linear min will be the expensive part (and therefore chooses the row_id-major index in one case), where actually the event_offset search is important and it's fine for the max to be linear on the tiny selected set, making the choice of the event_offset-only index the proper one.